### ISSUE TITLE
User with password containing a single quote cannot login

### ISSUE THREAD

**Issue by prometherion** [wp-cli/wp-cli#5089]
Installing WP using CLI, defining a password containing a single quote (`it'sweird`)
```
wp core install --url="http://domain.tld" \
--title="Innovative bottom-line flexibility" \
--admin_user="test.user" \
--admin_password="it'sweird" \
--admin_email="test.user@domain.tld" \
--path="/path/to/wp"
// WordPress installed successfully
```

Trying to login to `/wp-admin` with `test.user` / `it'sweird` doesn't work: so I decided to check the password using `wp shell`
```
$user = get_user_by('login', "test.user");
if ( $user && wp_check_password( "it'sweird", $user->data->user_pass, $user->ID) ) echo "seems ok";
// seems ok
```

After that, I decided to test with a password without single quote:
```
wp core install --url="http://domain.tld" \
--title="Innovative bottom-line flexibility" \
--admin_user="test.user" \
--admin_password="itsweird" \
--admin_email="test.user@domain.tld" \
--path="/path/to/wp"
// WordPress installed successfully
```

Checked password with `wp shell`: it's ok.
```
$user = get_user_by('login', "test.user");
if ( $user && wp_check_password( "itsweird", $user->data->user_pass, $user->ID) ) echo "seems ok";
// seems ok
```

Also the login using `wp-admin` is working as expected: editing password from UI using back `it'sweird`, logging out and logging in works without issue, so I suppose it's something not working with CLI...

**Comment by danielbachhuber** [comment]
Hey @prometherion,

I think I actually ran into this issue in a totally unrelated manner last week.

tldr: What I discovered is that passwords submitted via the web aren't unslashed before hashing. When you think your password is `it'sweird`, WordPress actually knows the password as `it\'sweird`.

**Comment by prometherion** [comment]
Thanks for the support @danielbachhuber!

However, I wanted to test your proposed workaround but even escaping on my own single quote, I still cannot log in.

**Comment by danielbachhuber** [comment]
> However, I wanted to test your proposed workaround but even escaping on my own single quote, I still cannot log in.

Can you share these reproduction steps?

**Comment by prometherion** [comment]
My bad, you're right... :)

Since I'm automating all the stuff with Go there were some escaping issues: works like a charm, thanks for the support @danielbachhuber!

**Comment by schlessera** [comment]
This is a known bug in WordPress that was never fixed.

I think we should compare the password to its slashed version and show a corresponding warning if they are not equal.

**Comment by danielbachhuber** [comment]
> I think we should compare the password to its slashed version and show a corresponding warning if they are not equal.

In WordPress core? I'm not sure where that would fit into WP-CLI...

**Comment by schlessera** [comment]
Huh? No, I meant if you do something like `--admin_password="it'sweird"` with WP-CLI, it should show a warning that your password will not be what you expect when using the graphical login page.

**Comment by amphrony** [comment]
Ran into similar issue, might be the same. Let me know if I should open new issue for this.

After doing `wp config create --dbpass="uh'oh" ...`, subsequent wp-cli commands fail with error:
`PHP Parse error:  syntax error, unexpected 'oh' (T_STRING), expecting ',' or ')' in phar:///usr/local/bin/wp/vendor/wp-cli/wp-cli/php/WP_CLI/Runner.php(1169) : eval()'d code on line 28`

I reproduced the error with this bash script:

```bash
#!/usr/bin/env bash

# Demonstrate the single quote bug in wp-cli

proj="wp_bugtest"
dbname="$proj"
dbuser="$proj"
dbpass="uh'oh"

cat <<END | mysql
CREATE USER '$dbuser'@'localhost' IDENTIFIED BY '${dbpass/\'/\\\'}';
GRANT ALL PRIVILEGES ON $dbname.* TO '$dbuser'@'localhost';
FLUSH PRIVILEGES;
CREATE DATABASE $dbname;
END

mkdir "$proj"
cd "$proj"

wp --version
wp core download --skip-content
wp config create --dbname="$dbname" --dbuser="$dbuser" --dbpass="$dbpass"
# From here, subsequent wp-cli commands fail.
wp help
wp core install
# PHP Parse error:  syntax error, unexpected 'oh' (T_STRING), expecting ',' or ')' in phar:///usr/local/bin/wp/vendor/wp-cli/wp-cli/php/WP_CLI/Runner.php(1169) : eval()'d code on line 28

# Cleanup
cd ..
rm -r "$proj"

cat <<END | mysql
DROP USER '$dbuser'@'localhost';
DROP DATABASE $dbname;
END
```

**Comment by schlessera** [comment]
@amphrony I think this is a different issue, related to how configuration files are being created. The above will create the following line in the `wp-config.php` file, as the input variables are not properly escaped:
```
define( 'DB_PASSWORD', 'uh'oh' );
```

Can you please open a separate bug report in the [wp-cli/config-command](https://github.com/wp-cli/config-command/issues/new?assignees=&labels=i%3A+bug&template=1-BUG_REPORT.md&title=)  repository?

**Comment by schlessera** [comment]
More details can be found at #5278, so closing this one.
